File: 3658.gcd-of-odd-and-even-sums/solution.kt
class Solution {
    fun gcdOfOddEvenSums(n: Int): Int {
        val sum = n * (2 * n + 1)
        val sumOdd = (sum - n) / 2
        val sumEven = sumOdd + n
        return gcd(sumOdd, sumEven)
    }

    tailrec fun gcd(a: Int, b: Int): Int =
        when {
            a == 0 -> b
            b == 0 -> a
            else -> gcd(b, a % b)
        }
}

File: 3658.gcd-of-odd-and-even-sums/solution.go
func gcdOfOddEvenSums(n int) int {
    sum := n * (2 * n + 1)
    sumOdd := (sum - n) / 2
    sumEven := sumOdd + n
    return gcd(sumOdd, sumEven)
}

func gcd(a, b int) int {
    for b != 0 {
        a, b = b, a%b
    }
    return a
}